typed_argparse/formatter.py

```diff
@@ -0,0 +1,23 @@
+import argparse
+from typing import Optional
+
+
+class DefaultHelpFormatter(argparse.HelpFormatter):
+    class Unformatted(str):
+        pass
+
+    def _get_help_string(self, action: argparse.Action) -> Optional[str]:
+        if (
+            not isinstance(action.help, DefaultHelpFormatter.Unformatted)
+            and action.help is not None
+            and action.default is not None
+            and action.default is not argparse.SUPPRESS
```

Comment:
There are a few things that I don't understand here yet:

1. It looks like we are going into the else branch if `help` is `DefaultHelpFormatter.Unformatted` or if it is `None`, right? Doesn't that mean that `DefaultHelpFormatter.Unformatted` is unnecessary? I.e., could the user achieve the same by setting `help` to `None` anyway?

2. What is the purpose of comparing `action.default` to `argparse.SUPPRESS`? In general it seems that `argparse.SUPPRESS` has multiple roles in argparse:
- It can be used as a value for the `help` argument to suppress the help text.
- It also can be used for the `argument_default` and `default` arguments (for the parser or add_argument). This usage has a special meaning that the field will be omitted from the `argparse.Namespace` instance.

The latter usage is problematic from the typed_argparse perspective, because we rely on arguments always being generated in the `Namespace` instance. Using `argparse.SUPPRESS` in the default position actually breaks argument parsing currently:

```py
import argparse
from typing import Optional

import typed_argparse as tap


class Args(tap.TypedArgs):
    foo: Optional[str] = tap.arg(default=argparse.SUPPRESS)


def run(args: Args):
    print(args)


if __name__ == "__main__":
    tap.Parser(Args).bind(run).run()
```

Fortunately it doesn't type check for other types than `str`, because its a string based sentinel value. I'll probably have to catch that, i.e., the default value `'==SUPPRESS=='` for strings is kinda forbidden :confused: 

This makes we wonder why to compare with `action.default` here though. I would have understood if there is a comparison like `action.help == argparse.SUPPRESS` but surprisingly that doesn't seem necessary - so the filtering actually happens internally in argparse? If so, shouldn't we remove `action.default is not argparse.SUPPRESS` from the condition, if we cannot support it in the `default` position anyway?

Reply:
Don't worry finding the time for this. There's no hurry.

1. `DefaultHelpFormatter.Unformatted` is a bit of a hack to support the `auto_default_help` option and is only used if `help` is not None (see https://github.com/MikkelSchubert/typed-argparse/blob/8a76abd4166a1bcbdd16554ce56851516bcb42be/typed_argparse/parser.py#L480). The problem is that `auto_default_help` isn't available to the `HelpFormatter` instance, so instead I convert the (not None) help string to an instance of a string subclass. This allows me to check whether or not the user set `auto_default_help=True` without breaking stuff in argparse, where `help` is expected to be a string. There may be a better way of doing this, but I couldn't figure one out.

2. I agree with all your points regarding `argparse.SUPPRESS`, but it needs to be handled here simply because argparse makes use of it even if we don't. I should probably add a comment about that. If you don't check for it in the `HelpFormatter` then you get output like this from (from an unmodified `examples/high_level_api/basic_usage.py`):

          -h, --help           show this help message and exit [default: ==SUPPRESS==]